File: examples/go/client_schema.go
package main

import (
	"fmt"
	"log"
	"os"

	ekodb "github.com/ekoDB/ekodb-client-go"
	"github.com/joho/godotenv"
)

func main() {
	// Load environment variables
	if err := godotenv.Load(); err != nil {
		log.Printf("Warning: .env file not found")
	}

	baseURL := os.Getenv("API_BASE_URL")
	if baseURL == "" {
		baseURL = "http://localhost:8080"
	}

	apiKey := os.Getenv("API_BASE_KEY")
	if apiKey == "" {
		apiKey = "a-test-api-key-from-ekodb"
	}

	// Create client
	client, err := ekodb.NewClient(baseURL, apiKey)
	if err != nil {
		log.Fatalf("Failed to create client: %v", err)
	}

	fmt.Println("=== Schema Management Examples ===\n")

	// Use unique collection names
	usersCollection := "schema_users_client_go"
	productsCollection := "schema_products_client_go"
	documentsCollection := "schema_documents_client_go"
	employeesCollection := "schema_employees_client_go"

	// Cleanup: Delete collections if they exist from previous runs
	client.DeleteCollection(usersCollection)
	client.DeleteCollection(productsCollection)
	client.DeleteCollection(documentsCollection)
	client.DeleteCollection(employeesCollection)

	// Example 1: Create a simple user schema
	fmt.Println("1. Creating user schema with basic fields:")
	userSchema := ekodb.NewSchemaBuilder().
		AddField("name", ekodb.NewFieldTypeSchemaBuilder("String").
			Required().
			Pattern("^[a-zA-Z ]+$").
			Build()).
		AddField("email", ekodb.NewFieldTypeSchemaBuilder("String").
			Required().
			Unique().
			Pattern("^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,}$").
			Build()).
		AddField("age", ekodb.NewFieldTypeSchemaBuilder("Integer").
			Range(0, 150).
			Build()).
		AddField("status", ekodb.NewFieldTypeSchemaBuilder("String").
			DefaultValue("active").
			Enums([]interface{}{"active", "inactive", "suspended"}).
			Build()).
		Build()

	if err := client.CreateCollection(usersCollection, userSchema); err != nil {
		log.Printf("Error: %v", err)
	} else {
		fmt.Println("✅ User schema created\n")
	}

	// Example 2: Schema with text index for full-text search
	fmt.Println("2. Creating product schema with text index:")
	productSchema := ekodb.NewSchemaBuilder().
		AddField("name", ekodb.NewFieldTypeSchemaBuilder("String").
			Required().
			TextIndex("english").
			Build()).
		AddField("description", ekodb.NewFieldTypeSchemaBuilder("String").
			TextIndex("english").
			Build()).
		AddField("price", ekodb.NewFieldTypeSchemaBuilder("Float").
			Required().
			Range(0, nil).
			Build()).
		AddField("category", ekodb.NewFieldTypeSchemaBuilder("String").
			Required().
			BTreeIndex().
			Build()).
		AddField("sku", ekodb.NewFieldTypeSchemaBuilder("String").
			Required().
			Unique().
			HashIndex().
			Build()).
		Build()

	if err := client.CreateCollection(productsCollection, productSchema); err != nil {
		log.Printf("Error: %v", err)
	} else {
		fmt.Println("✅ Product schema with indexes created\n")
	}

	// Example 3: Schema with vector index for semantic search
	fmt.Println("3. Creating document schema with vector index:")
	documentSchema := ekodb.NewSchemaBuilder().
		AddField("title", ekodb.NewFieldTypeSchemaBuilder("String").
			Required().
			TextIndex("english").
			Build()).
		AddField("content", ekodb.NewFieldTypeSchemaBuilder("String").
			Required().
			TextIndex("english").
			Build()).
		AddField("embedding", ekodb.NewFieldTypeSchemaBuilder("Array").
			VectorIndex(ekodb.VectorIndexHNSW, ekodb.DistanceMetricCosine, 16, 200).
			Build()).
		AddField("tags", ekodb.NewFieldTypeSchemaBuilder("Array").
			Build()).
		AddField("created_at", ekodb.NewFieldTypeSchemaBuilder("DateTime").
			Build()).
		Build()

	if err := client.CreateCollection(documentsCollection, documentSchema); err != nil {
		log.Printf("Error: %v", err)
	} else {
		fmt.Println("✅ Document schema with vector index created\n")
	}

	// Example 4: Get collection schema
	fmt.Println("4. Retrieving collection schema:")
	schema, err := client.GetSchema(usersCollection)
	if err != nil {
		log.Printf("Error: %v", err)
	} else {
		fmt.Printf("Schema fields: %d fields\n", len(schema.Fields))
		if schema.Version != nil {
			fmt.Printf("Schema version: %d\n", *schema.Version)
		}
	}
	fmt.Println()

	// Example 5: Get collection metadata
	fmt.Println("5. Retrieving collection metadata:")
	metadata, err := client.GetCollection(usersCollection)
	if err != nil {
		log.Printf("Error: %v", err)
	} else {
		fmt.Printf("Collection has %d fields\n", len(metadata.Collection.Fields))
	}
	fmt.Println()

	// Example 6: Complex schema with all constraint types
	fmt.Println("6. Creating employee schema with all constraint types:")
	employeeSchema := ekodb.NewSchemaBuilder().
		AddField("employee_id", ekodb.NewFieldTypeSchemaBuilder("String").
			Required().
			Unique().
			Pattern("^EMP-[0-9]{6}$").
			HashIndex().
			Build()).
		AddField("first_name", ekodb.NewFieldTypeSchemaBuilder("String").
			Required().
			Build()).
		AddField("last_name", ekodb.NewFieldTypeSchemaBuilder("String").
			Required().
			Build()).
		AddField("department", ekodb.NewFieldTypeSchemaBuilder("String").
			Required().
			Enums([]interface{}{"Engineering", "Sales", "Marketing", "HR", "Finance"}).
			BTreeIndex().
			Build()).
		AddField("salary", ekodb.NewFieldTypeSchemaBuilder("Float").
			Required().
			Range(30000, 500000).
			Build()).
		AddField("hire_date", ekodb.NewFieldTypeSchemaBuilder("DateTime").
			Required().
			Build()).
		AddField("is_active", ekodb.NewFieldTypeSchemaBuilder("Boolean").
			DefaultValue(true).
			Build()).
		AddField("bio", ekodb.NewFieldTypeSchemaBuilder("String").
			TextIndex("english").
			Build()).
		BypassRipple(false).
		Build()

	if err := client.CreateCollection(employeesCollection, employeeSchema); err != nil {
		log.Printf("Error: %v", err)
	} else {
		fmt.Println("✅ Employee schema with all constraints created\n")
	}

	fmt.Println("✅ Schema management examples completed!")
}

File: examples/typescript/client_schema.ts
/**
 * Schema Management Example
 *
 * Demonstrates schema creation and management:
 * - Creating collections with schemas
 * - Field types and constraints
 * - Indexes (text, vector, btree, hash)
 * - Validation rules
 */

import {
  EkoDBClient,
  SchemaBuilder,
  FieldTypeSchemaBuilder,
  VectorIndexAlgorithm,
  DistanceMetric,
} from "@ekodb/ekodb-client";
import * as dotenv from "dotenv";

dotenv.config();

async function main() {
  const client = new EkoDBClient(
    process.env.API_BASE_URL || "http://localhost:8080",
    process.env.API_BASE_KEY || "a-test-api-key-from-ekodb",
  );
  await client.init();

  console.log("=== Schema Management Examples ===\n");

  // Use unique collection names
  const usersCollection = "schema_users_client_ts";
  const productsCollection = "schema_products_client_ts";
  const documentsCollection = "schema_documents_client_ts";
  const employeesCollection = "schema_employees_client_ts";

  // Cleanup: Delete collections if they exist from previous runs
  try {
    await client.deleteCollection(usersCollection);
    await client.deleteCollection(productsCollection);
    await client.deleteCollection(documentsCollection);
    await client.deleteCollection(employeesCollection);
  } catch (e) {
    // Collections don't exist, that's fine
  }

  // Example 1: Create a simple user schema
  console.log("1. Creating user schema with basic fields:");
  const userSchema = new SchemaBuilder()
    .addField(
      "name",
      new FieldTypeSchemaBuilder("String").required().pattern("^[a-zA-Z ]+$"),
    )
    .addField(
      "email",
      new FieldTypeSchemaBuilder("String")
        .required()
        .unique()
        .pattern("^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,}$"),
    )
    .addField("age", new FieldTypeSchemaBuilder("Integer").range(0, 150))
    .addField(
      "status",
      new FieldTypeSchemaBuilder("String")
        .defaultValue("active")
        .enums(["active", "inactive", "suspended"]),
    )
    .build();

  await client.createCollection(usersCollection, userSchema);
  console.log("✅ User schema created\n");

  // Example 2: Schema with text index for full-text search
  console.log("2. Creating product schema with text index:");
  const productSchema = new SchemaBuilder()
    .addField(
      "name",
      new FieldTypeSchemaBuilder("String").required().textIndex("english"),
    )
    .addField(
      "description",
      new FieldTypeSchemaBuilder("String").textIndex("english"),
    )
    .addField(
      "price",
      new FieldTypeSchemaBuilder("Float").required().range(0, undefined),
    )
    .addField(
      "category",
      new FieldTypeSchemaBuilder("String").required().btreeIndex(),
    )
    .addField(
      "sku",
      new FieldTypeSchemaBuilder("String").required().unique().hashIndex(),
    )
    .build();

  await client.createCollection(productsCollection, productSchema);
  console.log("✅ Product schema with indexes created\n");

  // Example 3: Schema with vector index for semantic search
  console.log("3. Creating document schema with vector index:");
  const documentSchema = new SchemaBuilder()
    .addField(
      "title",
      new FieldTypeSchemaBuilder("String").required().textIndex("english"),
    )
    .addField(
      "content",
      new FieldTypeSchemaBuilder("String").required().textIndex("english"),
    )
    .addField(
      "embedding",
      new FieldTypeSchemaBuilder("Array").vectorIndex(
        VectorIndexAlgorithm.HNSW,
        DistanceMetric.Cosine,
        16,
        200,
      ),
    )
    .addField("tags", new FieldTypeSchemaBuilder("Array"))
    .addField(
      "created_at",
      new FieldTypeSchemaBuilder("DateTime").defaultValue(
        new Date().toISOString(),
      ),
    )
    .build();

  await client.createCollection(documentsCollection, documentSchema);
  console.log("✅ Document schema with vector index created\n");

  // Example 4: Get collection schema
  console.log("4. Retrieving collection schema:");
  const retrievedSchema = await client.getSchema(usersCollection);
  console.log("Schema fields:", Object.keys(retrievedSchema.fields));
  console.log("Schema version:", retrievedSchema.version);
  console.log();

  // Example 5: Get collection metadata
  console.log("5. Retrieving collection metadata:");
  const metadata = await client.getCollection(usersCollection);
  console.log(
    "Collection has",
    Object.keys(metadata.collection.fields).length,
    "fields",
  );
  console.log();

  // Example 6: Complex schema with all constraint types
  console.log("6. Creating employee schema with all constraint types:");
  const employeeSchema = new SchemaBuilder()
    .addField(
      "employee_id",
      new FieldTypeSchemaBuilder("String")
        .required()
        .unique()
        .pattern("^EMP-[0-9]{6}$")
        .hashIndex(),
    )
    .addField("first_name", new FieldTypeSchemaBuilder("String").required())
    .addField("last_name", new FieldTypeSchemaBuilder("String").required())
    .addField(
      "department",
      new FieldTypeSchemaBuilder("String")
        .required()
        .enums(["Engineering", "Sales", "Marketing", "HR", "Finance"])
        .btreeIndex(),
    )
    .addField(
      "salary",
      new FieldTypeSchemaBuilder("Float").required().range(30000, 500000),
    )
    .addField("hire_date", new FieldTypeSchemaBuilder("DateTime").required())
    .addField(
      "is_active",
      new FieldTypeSchemaBuilder("Boolean").defaultValue(true),
    )
    .addField("bio", new FieldTypeSchemaBuilder("String").textIndex("english"))
    .bypassRipple(false)
    .build();

  await client.createCollection(employeesCollection, employeeSchema);
  console.log("✅ Employee schema with all constraints created\n");

  console.log("✅ Schema management examples completed!");
}

main().catch((error) => {
  console.error("❌ Error:", error);
  process.exit(1);
});
